pkg/cloudprovider/provider/vsphere/provider.go

```diff
@@ -185,8 +208,17 @@ func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
 	}
 	defer session.Logout()
 
-	if _, err := session.Finder.Datastore(ctx, config.Datastore); err != nil {
-		return fmt.Errorf("failed to get datastore %s: %v", config.Datastore, err)
+	// If the datastore cluster is present it takes precedence over datastore.
+	if config.DatastoreCluster != "" && config.Datastore == "" {
```

Comment:
Since you point that out explicitly that datasotre cluster has a higher priority here, this code can be simpler by: 
```go
if config.DatastoreCluster == "" && config.Datastore == "" {
		return fmt.Errorf("at least one of datastore and datastore cluster should be specified: %v", err)
	}

if config.DatastoreCluster != "" {
	if _, err := session.Finder.DatastoreCluster(ctx, config.DatastoreCluster); err != nil {
		return fmt.Errorf("failed to get datastore cluster %s: %v", config.DatastoreCluster, err)
	}
} else {
	if _, err := session.Finder.Datastore(ctx, config.Datastore); err != nil {
		return fmt.Errorf("failed to get datastore %s: %v", config.Datastore, err)
	}
}
```
This way you have an early exist and you are not building condition trees.

Reply:
@moadqassem Actually the comment is outdated, finally I decided to have the two properties mutually exclusive, because it's misleading if both are specified. It's better to fail in such scenario. I'll update the comment.